### ISSUE TITLE
RPM preinstall fails when consul user does not exist

### ISSUE THREAD

**Issue by spaulg** [hashicorp/consul#8510]
#### Overview of the Issue

When installing the Consul RPM from the new yum repository, the install fails during the preinstall script if the consul user does not exist.

The script does create the user, but fails to create the home directory. Running the install again then succeeds as the user does not need to be created but the directory does not exist.

I also noticed, because the scripts use the --create-home flag, they'll be copying the /etc/skel directory, which contains other files such as .bash_logout, .bash_profile and .bashrc, as well as anything the admin might have also customised the machines install, and no doubt not intended for a config directory. I think it would be better to not create the directory via useradd. Instead create it manually after the user is created with something like: `sudo install -d -o consul -g consul -m 0700 /etc/consul.d`.

#### Reproduction Steps

Steps to reproduce this issue:

1. Ensure the Consul RPM is not installed, the consul user does not exist and the /etc/consul.d directory does not exist
1. Add the yum repo if not already added: `sudo yum-config-manager --add-repo https://rpm.releases.hashicorp.com/RHEL/hashicorp.repo`
1. Install Consul via yum: `sudo yum install -y consul`
1. View error

```
[vagrant@dp2 ~]$ getent passwd consul
[vagrant@dp2 ~]$ ls -lh /etc/consul.d/
ls: cannot access /etc/consul.d/: No such file or directory
[vagrant@dp2 ~]$ sudo yum install -y consul
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.vorboss.net
 * extras: centos.mirrors.nublue.co.uk
 * updates: mirrors.melbourne.co.uk
Resolving Dependencies
--> Running transaction check
---> Package consul.x86_64 0:1.8.3-1 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

==============================================================================================================================================================================================================================================================================================================================
 Package                                                                     Arch                                                                        Version                                                                         Repository                                                                      Size
==============================================================================================================================================================================================================================================================================================================================
Installing:
 consul                                                                      x86_64                                                                      1.8.3-1                                                                         hashicorp                                                                       33 M

Transaction Summary
==============================================================================================================================================================================================================================================================================================================================
Install  1 Package

Total download size: 33 M
Installed size: 108 M
Downloading packages:
consul-1.8.3-1.x86_64.rpm                                                                                                                                                                                                                                                                              |  33 MB  00:00:03     
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
useradd: cannot create directory /etc/consul.d
error: %pre(consul-1.8.3-1.x86_64) scriptlet failed, exit status 12
Error in PREIN scriptlet in rpm package consul-1.8.3-1.x86_64
  Verifying  : consul-1.8.3-1.x86_64                                                                                                                                                                                                                                                                                      1/1 

Failed:
  consul.x86_64 0:1.8.3-1                                                                                                                                                                                                                                                                                                     

Complete!
[vagrant@dp2 ~]$ 
[vagrant@dp2 ~]$ getent passwd consul
consul:x:995:990::/etc/consul.d:/bin/false
[vagrant@dp2 ~]$ ls -lh /etc/consul.d/
ls: cannot access /etc/consul.d/: No such file or directory
[vagrant@dp2 ~]$ 

```

### Operating system and Environment details

CentOS 7, running on VirtualBox with vagrant for x86_64.

[vagrant@dp2 ~]$ cat /etc/*-release
CentOS Linux release 7.8.2003 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.8.2003 (Core)
CentOS Linux release 7.8.2003 (Core)

**Comment by spaulg** [comment]
This appears to be due to SELinux. I'm guessing /etc/ for home directories aren't part of the default targeted policy, so the command fails.

```
[vagrant@dp2 ~]$ sudo setenforce 0
[vagrant@dp2 ~]$ sestatus
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   permissive
Mode from config file:          enforcing
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Max kernel policy version:      31
[vagrant@dp2 ~]$ sudo rpm -e consul
error: package consul is not installed
[vagrant@dp2 ~]$ sudo userdel consul
[vagrant@dp2 ~]$ sudo rm -rf /etc/consul.d/
[vagrant@dp2 ~]$ sudo yum install -y consul
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.vorboss.net
 * extras: centos.mirrors.nublue.co.uk
 * updates: mirrors.melbourne.co.uk
Resolving Dependencies
--> Running transaction check
---> Package consul.x86_64 0:1.8.3-1 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

==============================================================================================================================================================================================================================================================================================================================
 Package                                                                     Arch                                                                        Version                                                                         Repository                                                                      Size
==============================================================================================================================================================================================================================================================================================================================
Installing:
 consul                                                                      x86_64                                                                      1.8.3-1                                                                         hashicorp                                                                       33 M

Transaction Summary
==============================================================================================================================================================================================================================================================================================================================
Install  1 Package

Total download size: 33 M
Installed size: 108 M
Downloading packages:
consul-1.8.3-1.x86_64.rpm                                                                                                                                                                                                                                                                              |  33 MB  00:00:01     
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
Warning: RPMDB altered outside of yum.
  Installing : consul-1.8.3-1.x86_64                                                                                                                                                                                                                                                                                      1/1 
  Verifying  : consul-1.8.3-1.x86_64                                                                                                                                                                                                                                                                                      1/1 

Installed:
  consul.x86_64 0:1.8.3-1                                                                                                                                                                                                                                                                                                     

Complete!
[vagrant@dp2 ~]$ 
```

**Comment by kpenfound** [comment]
Thanks for reporting this @spaulg , I'm looking into it

**Comment by kpenfound** [comment]
This should be resolved now with `consul-1.8.3-2.x86_64.rpm`. The create-home wasn't necessary in the presinstall since `/etc/consul.d/consul.hcl` is provided by the package anyway

Tested in Vagrant centos/7 with `setenforce 1`:
```
Downloading packages:
consul-1.8.3-1.x86_64.rpm                                                                                                                                                                                                                              |  33 MB  00:00:05
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
useradd: cannot create directory /etc/consul.d
error: %pre(consul-1.8.3-1.x86_64) scriptlet failed, exit status 12
Error in PREIN scriptlet in rpm package consul-1.8.3-1.x86_64
  Verifying  : consul-1.8.3-1.x86_64                                                                                                                                                                                                                                      1/1

Failed:
  consul.x86_64 0:1.8.3-1

Complete!
```

```
Downloading packages:
consul-1.8.3-2.x86_64.rpm                                                                                                                                                                                                                              |  33 MB  00:00:05
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : consul-1.8.3-2.x86_64                                                                                                                                                                                                                                      1/1
  Verifying  : consul-1.8.3-2.x86_64                                                                                                                                                                                                                                      1/1

Installed:
  consul.x86_64 0:1.8.3-2

Complete!
```
